sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -127,13 +152,53 @@ case class TransformWithStateExec(
     mappedIterator
   }
 
+  private def processInitialStateRows(
+      keyRow: UnsafeRow,
+      initStateIter: Iterator[InternalRow]): Unit = {
+    val getKeyObj =
+      ObjectOperator.deserializeRowToObject(keyDeserializer, groupingAttributes)
+
+    val getStateValueObj =
+      ObjectOperator.deserializeRowToObject(initialStateDeserializer, initialStateDataAttrs)
+
+    val keyObj = getKeyObj(keyRow) // convert key to objects
+    ImplicitGroupingKeyTracker.setImplicitKey(keyObj)
+    val initStateObjIter = initStateIter.map(getStateValueObj.apply)
+
+    initStateObjIter.foreach { initState =>
+      statefulProcessor
+        .asInstanceOf[StatefulProcessorWithInitialState[Any, Any, Any, Any]]
+        .handleInitialState(keyObj, initState)
+    }
+    ImplicitGroupingKeyTracker.removeImplicitKey()
+  }
+
   private def processNewData(dataIter: Iterator[InternalRow]): Iterator[InternalRow] = {
     val groupedIter = GroupedIterator(dataIter, groupingAttributes, child.output)
     groupedIter.flatMap { case (keyRow, valueRowIter) =>
       val keyUnsafeRow = keyRow.asInstanceOf[UnsafeRow]
       handleInputRows(keyUnsafeRow, valueRowIter)
     }
   }
+// TODO double check this
+  private def processNewDataWithInitialState(
+      dataIter: Iterator[InternalRow],
+      initStateIter: Iterator[InternalRow]): Iterator[InternalRow] = {
+
+    val groupedChildDataIter = GroupedIterator(dataIter, groupingAttributes, child.output)
+    val groupedInitialStateIter =
+      GroupedIterator(initStateIter, initialStateGroupingAttrs, initialState.output)
+
+    // Create a CoGroupedIterator that will group the two iterators together for every key group.
+    new CoGroupedIterator(
+      groupedChildDataIter, groupedInitialStateIter, groupingAttributes).flatMap {
+      case (keyRow, valueRowIter, initialStateRowIter) =>
+        // TODO in design doc: trying to re-initialize state for the same
+        // grouping key will result in an error?
```

Comment:
yes - can we throw an error here if user tries to re-init state for same grouping key ?

Reply:
Not sure if I understand the scenario correctly, does it mean user tries to reassign state variable values in user defined function `handleInitialState()`?